Add topic 'Getting started with Tracing'. (#653)
tokio-rs/website: content/tokio/topics/tracing.md

trace the logic flow. `tracing` expands upon logging-style diagnostics by 
allowing libraries and applications to record structured events with additional 
information about *temporality* and *causality* — unlike a log message, a span 
in `tracing` has a beginning and end time, may be entered and exited by the 
flow of execution, and may exist within a nested tree of similar spans. In 
addition, `tracing` spans are *structured*, with the ability to record typed 
data as well as textual messages.

You can use `tracing` to:
- observe your application with Open Telemetry-supporting services
- debug your application with Tokio Console
- emit logs to `stdout` or `journald`
- profile where your application is spending time

[`tracing`]: https://docs.rs/tracing
[`tracing-subscriber`]: https://docs.rs/tracing-subscriber

# Setup

To begin, add [`tracing`] and [`tracing-subscriber`] as dependencies:

```toml
[dependencies]
tracing = "0.1.34"
tracing-subscriber = "0.3"
```

The [`tracing`] crate provides the API we will use to emit traces. The 
[`tracing-subscriber`] crate provides some basic utilities for forwarding those 

> traces to external listeners (e.g., `stdout`).

Author: Is `tracing-core` something that most folks interact with directly?

> traces to external listeners (e.g., `stdout`).

Reviewer: > Is tracing-core something that most folks interact with directly?

it isn't, but it is an important idea within tracing and explains some of the seemingly odder choices.

> traces to external listeners (e.g., `stdout`).

Reviewer 2: i think it ought to be _mentioned_ --- it's going to show up in your dependency tree even if you don't need to use its apis directly.

> traces to external listeners (e.g., `stdout`).

Reviewer 2: if this is intended as a beginner's document, i don't think it's particularly important to discuss tracing-core here. let's move on without adding it.

> # Subscribing to Traces

Author: I order the sections in this way because adding instrumentation is a task that can be tackled incrementally. If you install the subscriber first, you'll see the fruits of each addition of instrumentation.

> # Subscribing to Traces

Reviewer 2: okay, that's fine; i don't have a strong opinion about this.

register a tracing *subscriber*. Subscribers are types that process traces 
emitted by your application and its dependencies, and and can perform tasks 
such as computing metrics, monitoring for errors, and re-emitting traces to the 
outside world (e.g., `journald`, `stdout`, or an `open-telemetry` daemon).

In most circumstances, you should register your tracing subscriber as early as
possible in your `main` function. For instance, the [`FmtSubscriber`] type
provided by [`tracing-subscriber`] prints formatted traces and events to
`stdout`, can be registered like so:

```rust
#[tokio::main]
pub async fn main() -> mini_redis::Result<()> {
    // construct a subscriber that prints formatted traces to stdout
    let subscriber = tracing_subscriber::FmtSubscriber::new();
    // use that subscriber to process traces emitted after this point
    tracing::subscriber::set_global_default(subscriber)?;

    ...
}
```

[`FmtSubscriber`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/index.html

If you run your application now, you may see some trace events emitted by tokio,
but you will need to modify your own application to emit traces to get the most
out of `tracing`.

> 

Reviewer: +1. Layers should _absolutely_ be mentioned here.

# Emitting Traces

The easiest way to emit traces is with the [`instrument`] proc-macro annotation
provided by [`tracing`], which re-writes the bodies of functions to emit traces

> each time they are invoked.

Reviewer 2: i would probably _start_ by showing events, and then move on to also showing spans. it might even be worthwhile to show some sample output from the events and spans we add, so that the user knows what to expect --- for example, the default `FmtSubscriber` configuration doesn't print a log line when a span is entered/exited --- it only displays the current spans as context when an event is recorded. people have sometimes been surprised by this, so it would be nice if the tutorial explained that up front.

> each time they are invoked.

Reviewer: The other bit—it doesn't even technically emit _trace_ each time they're invoked—depending on the function being instrumented, it can either be a _new_ span or a _re-entered_ span, if it's an async function. While I see the utility of this guide, I think that a guide that provides slightly incorrect/imprecise documentation is a concern.

> each time they are invoked.

Reviewer 2: mini-redis _does_ have events in it, though, and even briefly discusses the event macro syntax: https://github.com/tokio-rs/mini-redis/blob/ebe4e1f33113d8e43a68199d04943b2581b33604/src/server.rs#L358-L367

we don't need to go into great detail on the syntax and different ways to use the macros; we could just refer to the reference-level documentation here: https://docs.rs/tracing/latest/tracing/#using-the-macros

however, i think this documentation is incomplete if we don't at least _mention_ the following:
- tracing has something called an "event" in it
- events represent moments in time while spans have starts and ends,
- events are recorded using function-like macros (link to docs, one or two examples)

it doesn't have to be a _lot_ of discussion of events, but we should at least say that they exist...

[`instrument`]: https://docs.rs/tracing/latest/tracing/attr.instrument.html

For instance, to trace the method in `mini-redis-server` that handles each
connection:

```rust
impl Handler {
    /// Process a single connection.
    #[instrument(
        level = "info",
        name = "Handler::run",
        skip(self),
        fields(

>             ?peer_addr = self.connection.peer_addr().unwrap(),

Reviewer: +1

information about *temporality* and *causality* — unlike a log message, a span 
in `tracing` has a beginning and end time, may be entered and exited by the 
flow of execution, and may exist within a nested tree of similar spans. In 
addition, `tracing` spans are *structured*, with the ability to record typed 
data as well as textual messages.

You can use `tracing` to:
- emit distributed traces to an [OpenTelemetry] collector
- debug your application with [Tokio Console]
- log to [`stdout`], [a log file] or [`journald`]
- [profile] where your application is spending time

[`tracing`]: https://docs.rs/tracing
[`tracing-subscriber`]: https://docs.rs/tracing-subscriber
[OpenTelemetry]: https://docs.rs/tracing-opentelemetry
[Tokio Console]: https://docs.rs/console-subscriber
[`stdout`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/index.html
[a log file]: https://docs.rs/tracing-appender/latest/tracing_appender/
[`journald`]: https://docs.rs/tracing-journald/latest/tracing_journald/
[profile]: https://docs.rs/tracing-timing/latest/tracing_timing/

# Setup

To begin, add [`tracing`] and [`tracing-subscriber`] as dependencies:

```toml
[dependencies]

> tracing = "0.1.34"

Reviewer: As per @hawkw's comment, I'll leave the suggestion that can be applied:
```suggestion
tracing = "0.1"
```

traces to external listeners (e.g., `stdout`).

# Subscribing to Traces

If you are authoring an executable (as opposed to a library), you will need to 
register a tracing *subscriber*. Subscribers are types that process traces 
emitted by your application and its dependencies, and and can perform tasks 
such as computing metrics, monitoring for errors, and re-emitting traces to the 
outside world (e.g., `journald`, `stdout`, or an `open-telemetry` daemon).

In most circumstances, you should register your tracing subscriber as early as
possible in your `main` function. For instance, the [`FmtSubscriber`] type
provided by [`tracing-subscriber`] prints formatted traces and events to
`stdout`, can be registered like so:

```rust
#[tokio::main]
pub async fn main() -> mini_redis::Result<()> {
    // construct a subscriber that prints formatted traces to stdout
    let subscriber = tracing_subscriber::FmtSubscriber::new();
    // use that subscriber to process traces emitted after this point
    tracing::subscriber::set_global_default(subscriber)?;

    ...
}
```

[`FmtSubscriber`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/index.html

> If you run your application now, you may see some trace events emitted by tokio,

Reviewer: ```suggestion
If you run your application now, you may see some trace events emitted by Tokio,
```

but you will need to modify your own application to emit traces to get the most
out of `tracing`.

# Emitting Traces

The easiest way to emit traces is with the [`instrument`] proc-macro annotation
provided by [`tracing`], which re-writes the bodies of functions to emit traces
each time they are invoked.

[`instrument`]: https://docs.rs/tracing/latest/tracing/attr.instrument.html

For instance, to trace the method in `mini-redis-server` that handles each
connection:

```rust
impl Handler {
    /// Process a single connection.
    #[instrument(

>         level = "info",

Reviewer: Eliza alluded [to this earlier](https://github.com/tokio-rs/website/pull/653/files#r866154130), but I wouldn't start with this example—we already default to `INFO` and this might contribute to overload. I strongly suggest starting simple. It can also be helpful to show the desugared version of this macro in terms of `span!`.

Reviewer 2: i'm fine with not showing the function-like `span!` macro in this documentation...it might be worth just including a link to the API docs that says like "for details on how spans are recorded, see here" or something.

        name = "Handler::run",
        skip(self),
        fields(
            ?peer_addr = self.connection.peer_addr().unwrap(),
        ),
    )]
    async fn run(&mut self) -> crate::Result<()> {
        ...
    }
}
```

> Now, `mini-redis-server` will emit a tracing span for each incoming connection

Reviewer: ```suggestion
`mini-redis-server` will now emit a `tracing` Span for each incoming connection
```

that:

> 1. has a priority [level] of `info` (the second-highest priority)

Reviewer: ```suggestion
1. has a verbosity [level] of `info` (the "middle ground" verbosity),
```

> 2. is named `Handler::run`

Reviewer: ```suggestion
2. is named `Handler::run`,
```

> 3. has some structured data associated with it:   
>     - `skip(self)` indicates that emitted traces should *not*
>     include the `Debug` representation of `Handler`
>     - `fields(...)` indicates that emitted traces *should* include
>     the `Debug` representation of the connection's `SocketAddr`, in a field
>     called `peer_addr`

Reviewer: ```suggestion
3. has some structured data associated with it.
    - `fields(...)` indicates that emitted span *should* include
    the `fmt::Debug` representation of the connection's `SocketAddr` in a field
    called `peer_addr`. 
    - `skip(self)` indicates that emitted span should *not* record `Hander`'s debug representation.
```

[level]: https://docs.rs/tracing/latest/tracing/struct.Level.html

> If you run your application, you now will see trace events emitted for each

Reviewer: ```suggestion
If you run your application, you now will see events decorated with their span's context emitted for each
```

# Subscribing to Traces

If you are authoring an executable (as opposed to a library), you will need to
register a tracing *subscriber*. Subscribers are types that process traces
emitted by your application and its dependencies, and and can perform tasks
such as computing metrics, monitoring for errors, and re-emitting traces to the
outside world (e.g., `journald`, `stdout`, or an `open-telemetry` daemon).

In most circumstances, you should register your tracing subscriber as early as
possible in your `main` function. For instance, the [`FmtSubscriber`] type
provided by [`tracing-subscriber`] prints formatted traces and events to
`stdout`, can be registered like so:

```rust
# mod mini_redis {
#       pub type Error = Box<dyn std::error::Error + Send + Sync>;
#       pub type Result<T> = std::result::Result<T, Error>;
# }
#[tokio::main]
pub async fn main() -> mini_redis::Result<()> {
    // construct a subscriber that prints formatted traces to stdout
    let subscriber = tracing_subscriber::FmtSubscriber::new();
    // use that subscriber to process traces emitted after this point
    tracing::subscriber::set_global_default(subscriber)?;
# /*
    ...
# */ Ok(())
}

> ```

Reviewer 3: The hashed sections here and at the end of the block are kind of throwing me off (partially because my brain is trying to understand what the hashes indicate in Rust (nothing right? I think you were intending to comment-out that code in a way) and also because the line right after is the annotation which does use a hash..

Do we really need the `mod mini_redis` section?

I'd maybe just write it as

```rust
#[tokio::main]
pub async fn main() -> mini_redis::Result<()> {
    // construct a subscriber that prints formatted traces to stdout
    let subscriber = tracing_subscriber::FmtSubscriber::new();
    // use that subscriber to process traces emitted after this point
    tracing::subscriber::set_global_default(subscriber)?;

    ... // the rest of main

    Ok(())
}
```

Author: Lines beginning with `#` are hidden from the rendered output, but processed when the code blocks are run as doc tests.

Reviewer 3: Oh I see, so the website won't show them? Gotcha -- I was just looking at the file directly as GitHub renders it. You can ignore my comments related to this then

If you run your application now, you may see some trace events emitted by Tokio,
but you will need to modify your own application to emit traces to get the most
out of `tracing`.

##  Subscriber Configuration

In the above example, we've configured [`FmtSubscriber`] with its default
configuration. However, `tracing-subscriber` also provides a number of ways to
configure the [`FmtSubscriber`]'s behavior, such as customizing the output
format, including additional information (such as thread IDs or source code
locations) in the logs, and writing the logs to somewhere other than `stdout`.

For example:
```rust
    // Start configuring a `fmt` subscriber
    let subscriber = tracing_subscriber::fmt()
        // Use a more compact, abbreviated log format
        .compact()
        // Display source code file paths
        .with_file(true)
        // Display source code line numbers
        .with_line_number(true)
        // Display the thread ID an event was recorded on
        .with_thread_ids(true)
        // Don't display the event's target (module path)
        .with_target(false)
        // Build the subscriber
        .finish();

> ```

Reviewer 3: Nit: Can we unindent the code?

interest. See [here][related-crates] for an (incomplete) list of additional
`Subscriber` implementations.

Finally, in some cases, it may be useful to combine multiple different ways of
recording traces together to build a single `Subscriber` that implements
multiple behaviors. For this purpose, the `tracing-subscriber` crate provides a
[`Layer`] trait that represents a component that may be composed together with
other `Layer`s to form a `Subscriber`. See [here][`Layer`] for details on using
`Layer`s.

[fmt-cfg]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/index.html#configuration
[related-crates]: https://docs.rs/tracing/latest/tracing/index.html#related-crates
[`Layer`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/index.html

# Emitting Spans and Events

The easiest way to emit spans is with the [`instrument`] proc-macro annotation
provided by [`tracing`], which re-writes the bodies of functions to emit spans
each time they are invoked.

[`instrument`]: https://docs.rs/tracing/latest/tracing/attr.instrument.html

For instance, to trace the method in `mini-redis-server` that handles each

> connection:

Reviewer 3: Since this is the first place that mini-redis is mentioned, I think we should have at least like a one-sentence explanation on what mini-redis is (could be inline or a footnote), if I was reading with no prior knowledge of tracing I'd be like "um, what is mini-redis-server??"

Reviewer 2: Yeah...maybe a link would be nice, too.

```rust
# struct Handler {
#     connection: tokio::net::TcpStream,

> # }

Reviewer 3: I'd maybe put this in a separate code block right above, or without the hashes and just add newlines, cause the hashes make the code look really weird when rendered https://github.com/tokio-rs/website/blob/00790803b06f7089fd5b212a7ac07c1f2b7a2b18/content/tokio/topics/tracing.md

# pub type Error = Box<dyn std::error::Error + Send + Sync>;

> # pub type Result<T> = std::result::Result<T, Error>;

Reviewer 3: I don't find these 2 lines particularly useful, they might just confuse the reader? Unless I'm missing something. Maybe we can just remove them.

use tracing::instrument;

impl Handler {
    /// Process a single connection.
    #[instrument(
        level = "info",
        name = "Handler::run",
        skip(self),
        fields(
            // `%` serializes the peer IP addr with `Display`
            peer_addr = %self.connection.peer_addr().unwrap()
        ),
    )]
    async fn run(&mut self) -> mini_redis::Result<()> {
# /*
        ...
# */ Ok::<_, _>(())

>     }

Reviewer 3: Same here, the hashes make the syntax highlighting look weird

}
```

`mini-redis-server` will now emit a `tracing` Span for each incoming connection

> that:

Reviewer 3: (This could/should probably just be a follow up PR)
It would be nice to have a sample of actual the log line in here

    the `fmt::Debug` representation of the connection's `SocketAddr` in a field
    called `peer_addr`.
    - `skip(self)` indicates that emitted span should *not* record `Hander`'s debug representation.

[level]: https://docs.rs/tracing/latest/tracing/struct.Level.html

You can also construct a [`Span`] manually by invoking the [`span!`] macro, or
any of its leveled shorthands ([`error_span!`], [`warn_span!`], [`info_span!`],
[`debug_span!`], [`trace_span!`]).

[`span!`]: https://docs.rs/tracing/*/tracing/macro.span.html
[`error_span!`]: https://docs.rs/tracing/*/tracing/macro.error_span.html
[`warn_span!`]: https://docs.rs/tracing/*/tracing/macro.warn_span.html
[`info_span!`]: https://docs.rs/tracing/*/tracing/macro.info_span.html
[`debug_span!`]: https://docs.rs/tracing/*/tracing/macro.debug_span.html
[`trace_span!`]: https://docs.rs/tracing/*/tracing/macro.trace_span.html

To emit events, invoke the [`event!`] macro, or any of its leveled shorthands
([`error!`], [`warn!`], [`info!`], [`debug!`], [`trace!`]). For instance, to
log that a client sent a malformed command:
```rust
# type Error = Box<dyn std::error::Error + Send + Sync>;

> # type Result<T> = std::result::Result<T, Error>;

Reviewer 3: Same comment as previous code block

# struct Command;
# impl Command {
#     fn from_frame<T>(frame: T) -> Result<()> {
#         Result::Ok(())
#     }
#     fn from_error<T>(err: T) {}
# }

> # let frame = ();

Reviewer 3: Some comment as previous code block

title: Getting started with Tracing
---

The [`tracing`] crate is a framework for instrumenting Rust programs to collect
structured, event-based diagnostic information.

In asynchronous systems like Tokio, interpreting traditional log messages can
often be quite challenging. Since individual tasks are multiplexed on the same
thread, associated events and log lines are intermixed making it difficult to
trace the logic flow. `tracing` expands upon logging-style diagnostics by
allowing libraries and applications to record structured events with additional
information about *temporality* and *causality* — unlike a log message, a `Span`
in `tracing` has a beginning and end time, may be entered and exited by the
flow of execution, and may exist within a nested tree of similar spans. For

> representing things that occur in a moment-in-time, `tracing` provides the

Reviewer 2: I don't think i would hyphenate this:
```suggestion
representing things that occur at a single moment in time, `tracing` provides the
```

> information about *temporality* and *causality* — unlike a log message, a `Span`
> in `tracing` has a beginning and end time, may be entered and exited by the
> flow of execution, and may exist within a nested tree of similar spans. For
> representing things that occur in a moment-in-time, `tracing` provides the
> complementary concept of *events*. Both `Span`s and `Event`s are *structured*,
> with the ability to record typed data as well as textual messages.

Reviewer 2: i think we should probably turn these into docs links?
```suggestion
information about *temporality* and *causality* — unlike a log message, a [`Span`]
in `tracing` has a beginning and end time, may be entered and exited by the
flow of execution, and may exist within a nested tree of similar spans. For
representing things that occur in a moment-in-time, `tracing` provides the
complementary concept of [*events*][`Event`]. Both [`Span`s] and [`Event`]s
are *structured*, with the ability to record typed data as well as textual messages.

[`Span`]: https://docs.rs/tracing/latest/tracing/#spans
[`Event`]: https://docs.rs/tracing/latest/tracing/#events
```

If you run your application now, you may see some trace events emitted by tokio,
but you will need to modify your own application to emit traces to get the most
out of `tracing`.

> ##  Subscriber Configuration

Reviewer 2: nit: i think there should be an extra line of whitespace here --- i think that was my fault
```suggestion
out of `tracing`.

##  Subscriber Configuration
```

        name = "Handler::run",
        skip(self),
        fields(

>             peer_addr = &self.connection.peer_addr().unwrap().to_string()[..]

Reviewer 2: hmm i don't love `to_string()`ing this here...can we just record the field using `Display` instead?
```suggestion
            peer_addr = %self.connection.peer_addr().unwrap()
```

we would need to change the text on line 181 to say that it's recorded using `Display` instead of `Debug` and that `%` means `Display`, but that text was already inaccurate...
